src/assessments/seo/keywordDensityAssessment.js

```diff
@@ -1,148 +1,265 @@
-		%3$s expands to the maximum keyword density percentage, %4$s expands to a link to a Yoast.com article
-		about keyword density, %5$s expands to the anchor end tag. */
-		text = i18n.dngettext(
-			"js-text-analysis",
-			"The %4$skeyword density%5$s is %1$s, which is over the advised %3$s maximum; the focus keyword was found %2$d time.",
-			"The %4$skeyword density%5$s is %1$s, which is over the advised %3$s maximum; the focus keyword was found %2$d times.",
-			keywordCount
-		);
+		this._keywordCount = researcher.getResearch( "keywordCount" ).count;
 
-		text = i18n.sprintf( text, keywordDensityPercentage, keywordCount, max, url, "</a>" );
-	}
+		this._keywordDensity = researcher.getResearch( "getKeywordDensity" );
 
-	if ( inRangeStartEndInclusive( roundedKeywordDensity, 0.5, 2.5 ) ) {
-		score = 9;
+		const calculatedScore = this.calculateResult( i18n );
+		assessmentResult.setScore( calculatedScore.score );
+		assessmentResult.setText( calculatedScore.resultText );
+		assessmentResult.setHasMarks( this._keywordCount > 0 );
 
-		/* Translators:	%1$s expands to the keyword density percentage,	%2$d expands to the keyword count,
-		%3$s expands to a link to a Yoast.com article about keyword density,
-		%4$s expands to the anchor end tag. */
-		text = i18n.dngettext(
-			"js-text-analysis",
-			"The %3$skeyword density%4$s is %1$s, which is great; the focus keyword was found %2$d time.",
-			"The %3$skeyword density%4$s is %1$s, which is great; the focus keyword was found %2$d times.",
-			keywordCount
-		);
+		return assessmentResult;
+	}
 
-		text = i18n.sprintf( text, keywordDensityPercentage, keywordCount, url, "</a>" );
+	/**
+	 * Checks whether there are no keyword matches in the text.
+	 *
+	 * @returns {boolean} Returns true if the keyword count is 0.
+	 */
+	hasNoMatches() {
+		return this._keywordCount === 0;
 	}
 
-	if ( roundedKeywordDensity < 0.5 && keywordCount === 0 ) {
-		score = 4;
+	/**
+	 * Checks whether there are too few keyword matches in the text. One keyword match is always considered
+	 * as bad, regardless of the density.
+	 *
+	 * @returns {boolean} Returns true if the rounded keyword density is between 0 and the recommended minimum
+	 * or if there there is only 1 keyword match (regardless of the density).
+	 */
+	hasTooFewMatches() {
+		return inRangeStartInclusive( this._keywordDensity, 0, this._config.minimum );
+	}
 
-		/* Translators:	%1$s expands to the keyword density percentage,	%2$d expands to the keyword count,
-		%3$s expands to a link to a Yoast.com article about keyword density,
-		%4$s expands to the anchor end tag. */
-		text = i18n.dgettext(
-			"js-text-analysis",
-			"The %3$skeyword density%4$s is %1$s, which is too low; the focus keyword was found %2$d times."
-		);
+	/**
+	 * Checks whether there is a good number of keyword matches in the text. Two keyword matches are always
+	 * considered as good, regardless of the density.
```

Comment:
The description needs to be changed - `Two keyword matches are always considered as good, regardless of the density.` only holds true for recalibration.